examples/twittermap/web/public/javascripts/lib/webgl_point_layer.js

```diff
@@ -0,0 +1,361 @@
+        gl.enableVertexAttribArray(this._programs[1].indexLoc);
+
+        this.render();
+    },
+
+
+    _initTextures: function() {
+        var gl = this._gl;
+        if ( gl == null ) return;
+
+        var canvas = this.getCanvas();
+        // console.log(canvas.width + ' x ' + canvas.height);
+
+        this._fbTexture = gl.createTexture();
+        gl.bindTexture(gl.TEXTURE_2D, this._fbTexture);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, canvas.width, canvas.height, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+
+        this._fb = gl.createFramebuffer();
+        gl.bindFramebuffer(gl.FRAMEBUFFER, this._fb);
+        gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this._fbTexture, 0);
+        gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+    },
+
+
+    _reset: function() {
+        L.CanvasLayer.prototype._reset.call(this);
+
+        var gl = this._gl;
+        if ( gl == null ) return;
+
+        var canvas = this.getCanvas();
+        var fbTexture = this._fbTexture;
+
+        gl.bindTexture(gl.TEXTURE_2D, fbTexture);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, canvas.width, canvas.height, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+    },
+
+
+    _mousemove: function(e) {
+        var canvas = this.getCanvas();
+        var rect = canvas.getBoundingClientRect();
+        this.setCurrentPoint(Math.floor(e.originalEvent.clientX - rect.left + 0.5),
+                             Math.floor(e.originalEvent.clientY - rect.top + 0.5));
+    },
+
+
+    render: function() {
+        var gl = this._gl, map = this._map;
+        if ( gl == null || map == null ) return;
+
+        var canvas = this.getCanvas();
+        var pixelsToWebGLMatrix = new Float32Array(16);
+        var mapMatrix = new Float32Array(16);
+
+        pixelsToWebGLMatrix.set([2 / canvas.width, 0, 0, 0, 0, -2 / canvas.height, 0, 0, 0, 0, 0, 0, -1, 1, 0, 1]);
+        var pointSize = Math.max(map.getZoom() - 4.0, this._pointSize);
+        // var pointSize = Math.max(map.getZoom() - 4.0, 1.0);
+        console.log("map.getZoom() = " + map.getZoom() + ", pointSize = " + pointSize);
+        mapMatrix.set(pixelsToWebGLMatrix);
+        var bounds = map.getBounds();
+        var topLeft = new L.LatLng(bounds.getNorth(), bounds.getWest());
+        var offset = this._LatLongToPixel_XY(topLeft.lat, topLeft.lng);
+        var scale = Math.pow(2, map.getZoom());
+        this._scaleMatrix(mapMatrix, scale, scale);
+        this._translateMatrix(mapMatrix, -offset.x, -offset.y);
+
+        gl.viewport(0, 0, canvas.width, canvas.height);
+
+        // Pass 1
```

Comment:
Can you share more details on what the pass 1 and 2 do and why?

Reply:
For these detailed questions, maybe we need a F2F meeting with Prof. Shuang Zhao to go through the code together. @chenlica What you think?